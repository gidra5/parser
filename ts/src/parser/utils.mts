import { err, ok, Result, none } from "../types.mjs";
import {
  isOperator,
  Operator,
  OperatorRegistry,
  Parser,
  ParserRecovery,
  Precedence,
  Error,
} from "./types.mjs";

export class ParsingHandler<T, U> {
  private resetIndex: number;
  private index: number;
  constructor(
    private src: T[],
    options?: {
      index?: number;
      resetIndex?: number;
    }
  ) {
    this.index = options?.index ?? 0;
    this.resetIndex = options?.resetIndex ?? this.index;
  }

  skip<A extends T[]>(values: A) {
    const isNext = this.peekIs(values);
    if (isNext) this.advance();
    return isNext;
  }
  skipAll<A extends T[]>(values: A) {
    while (this.skip(values)) {}
  }

  peekIs<A extends T[]>(values: A) {
    return values.includes(this.peek() as T);
  }

  peek(step = 0): T | undefined {
    return this.src[this.index + step];
  }

  next(step = 1): T | undefined {
    const item = this.src[this.index];
    this.advance(step);
    return item;
  }

  advance(step = 1) {
    this.index += step;
  }

  setIndex() {
    this.resetIndex = this.index;
  }

  reset() {
    this.index = this.resetIndex;
  }

  *parse<E>(
    ...args:
      | [handler: Parser<T, U, Error>]
      | [handler: Parser<T, U, E>, recover: ParserRecovery<T, U, E>]
  ) {
    const [handler, recover] = args;
    while (this.peek()) {
      // const start = this.index;
      const item = handler(this);

      if (item.type === "err") {
        let error = item.err;

        if (recover) {
          error = recover(this, item.err as E);
          this.setIndex();
        }

        if (error === Error.SKIP) continue;
        // yield { start, end: this.index, item: err<U, Error>(error as Error) };
        yield err<U, Error>(error as Error);
      } else {
        this.setIndex();
        // yield { start, end: this.index, item: ok<U, Error>(item.value) };
        yield ok<U, Error>(item.value);
      }
    }
  }
}

export const transpose = <T,>(x: Result<T, Error>[]): [T[], Error[]] => {
  const operands = [] as T[];
  const errors = [] as Error[];

  for (const it of x) {
    if (it.type === "err") errors.push(it.err);
    else operands.push(it.value);
  }

  return [operands, errors];
};

export const DEFAULT_PRECEDENCE = [none(), none()] as Precedence;
/**
 *
 * @param operatorRegistry
 * @returns left is `less`\\`greater`\\`equal` to right, or can't be compared
 */
export const compareOperators =
  (operatorRegistry: OperatorRegistry) =>
  (left?: Operator, right?: Operator): boolean => {
    if (isOperator(left)) console.log("c3", operatorRegistry.get(left.item));
    if (isOperator(right)) console.log("c4", operatorRegistry.get(right.item));

    const [, _leftBP] = isOperator(left)
      ? operatorRegistry.get(left.item).precedence ?? DEFAULT_PRECEDENCE
      : DEFAULT_PRECEDENCE;
    const [_rightBP] = isOperator(right)
      ? operatorRegistry.get(right.item).precedence ?? DEFAULT_PRECEDENCE
      : DEFAULT_PRECEDENCE;

    if (_rightBP.type === "none") return false;
    if (_leftBP.type === "none") return true;

    return _leftBP.value < _rightBP.value;
  };
