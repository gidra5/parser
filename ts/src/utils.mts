export function* range(
  start: number,
  end: number
): Generator<number, number, unknown> {
  while (start < end) {
    yield start++;
  }
  return start;
}

export const rangeFromZero = (
  end: number
): Generator<number, number, unknown> => range(0, end);

export function* iter<T>(x: Array<T>): Generator<T, void, unknown> {
  for (const item of x) {
    yield item;
  }
}

export function* map<T, U>(
  gen: IterableIterator<T>,
  map: (x: T) => U
): Generator<U, void, unknown> {
  for (const item of gen) {
    yield map(item);
  }
}

export function* enumerate<T>(
  gen: IterableIterator<T>
): Generator<[item: T, i: number], void, unknown> {
  let count = 0;
  for (const item of gen) {
    yield [item, count++] as [item: T, i: number];
  }
}

export function* repeat<T>(x: T, count: number): Generator<T, void, unknown> {
  for (let i = 0; i < count; ++i) yield x;
}

export function* product<T, U>(
  gen1: IterableIterator<T>,
  gen2: IterableIterator<U>
): Generator<[first: T, second: U], void, unknown> {
  const buffer = [...gen2];
  for (const item1 of gen1) {
    for (const item2 of buffer) {
      yield [item1, item2];
    }
  }
}

export function* groupBy<
  T extends unknown,
  U extends unknown,
  P extends (input: T[], index: number) => [U, number]
>(str: T[], parse: P) {
  let index = 0;

  while (str[index] !== "") {
    const [token, nextIndex] = parse(str, index);
    index = nextIndex;
    yield token;
  }
}
