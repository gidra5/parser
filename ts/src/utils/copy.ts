import { Iterator } from "iterator-js";

export const CopySymbol = Symbol("Copy");

export interface Copyable<T> {
  [CopySymbol](): T;
}

export const copy = <T>(x: T | Copyable<T>): T => {
  if (!x) return x;
  if (typeof x === "number") return x;
  if (typeof x === "string") return x;
  if (typeof x === "boolean") return x;
  if (typeof x === "bigint") return x;
  if (Array.isArray(x)) return x.map(copy) as T;
  if (x instanceof Set) return new Set([...x].map(copy)) as T;
  if (x instanceof Map) return new Map([...x].map(([k, v]) => [copy(k), copy(v)])) as T;

  if (typeof (x as any)[CopySymbol] === "function") {
    return (x as any)[CopySymbol]();
  }

  return Iterator.iterEntries(x as any)
    .map(([k, v]) => [k, copy(v)] as [string, any])
    .toObject() as T;
};
