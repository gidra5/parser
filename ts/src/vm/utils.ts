export function signExtend(x: number, bitCount: number) {
  const m = 1 << (bitCount - 1);
  x &= (1 << bitCount) - 1;
  return (x ^ m) - m;
}

export function putBuf(data: number[]) {
  process.stdout.write(Buffer.from(data).toString("utf8"));
}
