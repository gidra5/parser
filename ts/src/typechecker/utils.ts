import { AbstractSyntaxTree } from "../parser/ast";
import { Type } from "./types";

// check if value is assignable to variable
export const assignable = (valueType: Type, variableType: Type): boolean => {
  return false;
};

export const chechValue = (ast: AbstractSyntaxTree, type: Type): boolean => {
  return false;
};
