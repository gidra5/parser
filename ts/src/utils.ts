import { Token } from "./parser/types";

export function assert(condition: any, msg?: string): asserts condition {
  if (condition) return;
  throw new Error(msg ? `Assertion failed: ${msg}` : "Assertion failed");
}

export const omit = <T extends {}, K extends string>(
  obj: T,
  keys: K[]
): Omit<T, K> => {
  return Object.fromEntries(
    Object.entries(obj).filter(([key]) => !keys.includes(key as K))
  ) as Omit<T, K>;
};

export const pick = <T extends {}, K extends keyof T>(
  obj: T,
  keys: K[]
): Pick<T, K> => {
  return Object.fromEntries(
    Object.entries(obj).filter(([key]) => keys.includes(key as K))
  ) as Pick<T, K>;
};

export const clamp = (x: number, min: number, max: number) =>
  Math.min(Math.max(x, min), max);

export const isEqual = (a, b) => {
  if (a === b) return true;
  if (a === null || b === null) return false;
  if (typeof a !== "object" || typeof b !== "object") return false;

  if (Array.isArray(a) && Array.isArray(b)) {
    if (a.length !== b.length) return false;
    for (let i = 0; i < a.length; i++) {
      if (!isEqual(a[i], b[i])) return false;
    }
    return true;
  }

  if (Object.keys(a).length !== Object.keys(b).length) return false;
  for (const key in a) {
    if (!(key in b)) return false;
    if (!isEqual(a[key], b[key])) return false;
  }

  return true;
};
