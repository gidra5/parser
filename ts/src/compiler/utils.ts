import { Iterator } from "iterator-js";
import { disassemble } from "../vm/handlers";
import { CodeChunk, chunkToByteCode } from "./chunks";
import { RegisterState } from "./registers";

export const chunkToString = (chunk: CodeChunk): string => {
  return disassemble(chunkToByteCode([], [])(chunk, 0));
};
export const registerStateToObject = <T>(state: RegisterState<T>["state"]) => {
  return Iterator.iter(state)
    .enumerate()
    .flatMap(([x, i]) => (x ? [{ reg: i, ...x }] : []))
    .map<[number, any]>(({ reg, ...rest }) => [reg, rest])
    .toObject();
};
