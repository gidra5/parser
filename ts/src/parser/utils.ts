import {
  ParsingContext,
  Scope,
  defaultParsingContext,
  parse,
  parseExpr,
} from ".";
import { Iterator } from "iterator-js";
import { parseTokens } from "./tokens";
import { DefaultVisitor, Tree, Visitor } from "../tree";
import { isEqual } from "../utils";
import { AbstractSyntaxTree } from "./ast";
import { ConsumeParsingResult, ParsingResult, TokenParser } from "./types";

type TemplateValues = AbstractSyntaxTree[] | Record<string, AbstractSyntaxTree>;
export const template = (tree: AbstractSyntaxTree, values: TemplateValues) => {
  const visitor: Visitor<Tree, AbstractSyntaxTree> = new Visitor<
    Tree,
    AbstractSyntaxTree
  >({
    [DefaultVisitor]: (tree) => {
      const children = visitor.visitChildren(tree).toArray();
      return { ...tree, children };
    },
    placeholder: (tree) => {
      return (values as any[]).pop() ?? tree;
    },
    name: (tree) => {
      return values[tree.value as string] ?? tree;
    },
  });
  return visitor.visitNode(tree);
};

export const match = (
  tree: AbstractSyntaxTree,
  pattern: AbstractSyntaxTree,
  matches: Record<string, AbstractSyntaxTree> = {}
) => {
  type ReturnType = [boolean, Record<string, AbstractSyntaxTree>];
  const visitor: Visitor<ReturnType, AbstractSyntaxTree> = new Visitor({
    [DefaultVisitor]: (pattern) => {
      if (tree.name !== pattern.name) return [false, matches];
      if (tree.children.length !== pattern.children.length)
        return [false, matches];
      return Iterator.iter(tree.children)
        .zip(pattern.children)
        .reduce<ReturnType>(
          (acc, args) => {
            const [result, matches] = acc;
            const [nextResult, newMatches] = match(...args, matches);
            return [result && nextResult, newMatches];
          },
          [true, matches]
        );
    },
    placeholder: () => [true, matches] as ReturnType,
    name: (pattern: AbstractSyntaxTree): ReturnType => {
      const name = pattern.value as string;
      if (name in matches) {
        return [isEqual(tree, matches[name]), matches];
      }
      return [true, { ...matches, [name]: tree }];
    },
  });
  return visitor.visitNode(pattern);
};

export const parseString = (src: string, scope: Scope = {}) => {
  const [tokens] = parseTokens(src);
  const context = defaultParsingContext();
  context.scope = { ...context.scope, ...scope };
  const result = parseExpr(context)(tokens).slice(1, 3);
  return result as ConsumeParsingResult<AbstractSyntaxTree>;
};

export const templateString = (templateStr: string, values: TemplateValues) => {
  const [parsed] = parseString(templateStr);
  return template(parsed, values);
};

export const matchString = (
  tree: AbstractSyntaxTree,
  pattern: string,
  matches: Record<string, AbstractSyntaxTree> = {}
) => {
  const [patternParsed] = parseString(pattern);
  return match(tree, patternParsed, matches);
};

export const matchTokens =
  (...tokens: string[]): TokenParser<boolean> =>
  (src, i = 0) => {
    const found = tokens.includes(src[i]?.src);
    console.log("matchTokens", tokens, i, src[i]?.src, found);

    if (found) return [i + 1, true, []];
    return [i, false, []];
  };

export const mapParsingResult = <T, const U>(
  [index, result, errors]: ParsingResult<T>,
  fn: (result: T) => U
): ParsingResult<U> => [index, fn(result), errors];

export const mapParserResult =
  <T, const U>(parser: TokenParser<T>, fn: (result: T) => U): TokenParser<U> =>
  (src, i) =>
    mapParsingResult(parser(src, i), fn);

export const matchSeparators =
  (...separators: string[][]) =>
  (context: ParsingContext) => {
    const index = context.groupNodes?.length ?? 0;
    console.log("separators", separators, index, separators[index]);

    const separator = separators[index] ?? [];
    const isLast = index === separators.length - 1;
    return mapParserResult(matchTokens(...separator), (matched) =>
      matched ? (isLast ? "done" : "match") : "noMatch"
    );
  };
