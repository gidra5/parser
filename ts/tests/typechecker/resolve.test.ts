import { describe, expect } from "vitest";
import { it, fc, test } from "@fast-check/vitest";
import { exampleScopeTestCase, treeScopeTestCase } from "./utils";

describe("expressions", () => {
  describe("function expressions", () => {
    test("function multiple params", () => {
      const src = `fn x y -> x y`;
      treeScopeTestCase(src);
    });

    test("function", () => {
      const src = `x -> x`;
      treeScopeTestCase(src);
    });

    test("function with placeholder arg", () => {
      const src = `_ -> #0`;
      treeScopeTestCase(src);
    });

    test("function with no arg", () => {
      const src = `fn -> #0`;
      treeScopeTestCase(src);
    });

    test("function with name shadowing", () => {
      const src = `fn a -> fn a -> a`;
      treeScopeTestCase(src);
    });

    test("function with shadowed name access", () => {
      const src = `fn a -> fn a -> #a`;
      treeScopeTestCase(src);
    });

    test("function with deep shadowed name access", () => {
      const src = `fn a -> fn a -> fn a -> ##a`;
      treeScopeTestCase(src);
    });
  });

  describe("pattern matching", () => {
    test("in function parameters", () => {
      const src = `(x, y) -> x + y`;
      treeScopeTestCase(src);
    });

    test.todo("with 'is' operator", () => {
      const src = `x is (a, b) and a == b + 1`;
      treeScopeTestCase(src);
    });

    test.todo("with 'is' operator with equality", () => {
      const src = `x is (a, a, b) and a == b + 1`;
      treeScopeTestCase(src);
    });

    test.todo("with placeholder", () => {
      const src = `x is (_, b) and a == b + 1`;
      treeScopeTestCase(src);
    });

    test.todo("with variable value", () => {
      const src = `x is (^a, b) and a == b + 1`;
      treeScopeTestCase(src);
    });

    test.todo("with rest value", () => {
      const src = `x is (a, ...b) and a == b + 1`;
      treeScopeTestCase(src);
    });

    test.todo("with rest value first", () => {
      const src = `x is (...b, a) and a == b + 1`;
      treeScopeTestCase(src);
    });

    test.todo("binding visible in scope where it is true", () => {
      const src = `x is (a, b) and a == b + 1`;
      treeScopeTestCase(src);
    });
  });

  describe.todo("structured programming", () => {
    test("if-then", () => {
      const src = `if true: 123`;
      treeScopeTestCase(src);
    });

    test("if-then newline", () => {
      const src = `if true\n 123`;
      treeScopeTestCase(src);
    });

    test("if-then-else", () => {
      const src = `if true: 123 else 456`;
      treeScopeTestCase(src);
    });

    test("if-then-elseif-then-else", () => {
      const src = `if true: 123 else if false: 789 else 456`;
      treeScopeTestCase(src);
    });

    test("if-then newline-else", () => {
      const src = `if true\n 123 else 456`;
      treeScopeTestCase(src);
    });

    test("if-then newline-else newline", () => {
      const src = `if true\n 123 else\n 456`;
      treeScopeTestCase(src);
    });

    test("block", () => {
      const src = `{ 123 }`;
      treeScopeTestCase(src);
    });

    test("for loop", () => {
      const src = `for x in [1, 2, 3]: x`;
      treeScopeTestCase(src);
    });

    test("for loop newline", () => {
      const src = `for x in [1, 2, 3]\n x`;
      treeScopeTestCase(src);
    });

    test("while loop", () => {
      const src = `while true: 123`;
      treeScopeTestCase(src);
    });

    test("while loop break", () => {
      const src = `while true: break _`;
      treeScopeTestCase(src);
    });

    test("while loop break value", () => {
      const src = `while true: break 1`;
      treeScopeTestCase(src);
    });

    test("while loop continue", () => {
      const src = `while true: continue _`;
      treeScopeTestCase(src);
    });

    test("while loop continue value", () => {
      const src = `while true: continue 1`;
      treeScopeTestCase(src);
    });

    test("labeled expression", () => {
      const src = `label: 123`;
      treeScopeTestCase(src);
    });

    test("expression-label", () => {
      const src = `123+456: 789`;
      treeScopeTestCase(src);
    });

    test("return", () => {
      const src = `() -> { return 123 }`;
      treeScopeTestCase(src);
    });

    test("block variable declaration", () => {
      const src = `{ x := 123 }`;
      treeScopeTestCase(src);
    });

    test("block mutable variable declaration", () => {
      const src = `{ mut x := 123 }`;
      treeScopeTestCase(src);
    });

    test("block variable assignment", () => {
      const src = `{ x = 123 }`;
      treeScopeTestCase(src);
    });

    test("block pattern matching", () => {
      const src = `{ x, y = 123, 456 }`;
      treeScopeTestCase(src);
    });
  });
});

describe.todo("programs", () => {
  describe("script", () => {
    test("use", () => {
      const src = `use "a" as b`;
      treeScopeTestCase(src);
    });
    test("use with", () => {
      const src = `use "a" as b with x`;
      treeScopeTestCase(src);
    });
    test("export", () => {
      const src = `export x`;
      treeScopeTestCase(src);
    });
    test("export as", () => {
      const src = `export x as y`;
      treeScopeTestCase(src);
    });

    test("variable use", () => {
      const src = `x := 123;  use "a" as b`;
      treeScopeTestCase(src);
    });
    test("variable use with", () => {
      const src = `x := 123;  use "a" as b with x`;
      treeScopeTestCase(src);
    });
    test("variable export", () => {
      const src = `x := 123; export x`;
      treeScopeTestCase(src);
    });
    test("variable export as", () => {
      const src = `x := 123; export x as y`;
      treeScopeTestCase(src);
    });
  });

  describe("module", () => {
    test("import", () => {
      const src = `import "a" as b`;
      treeScopeTestCase(src);
    });
    test("import with", () => {
      const src = `import "a" as b with c`;
      treeScopeTestCase(src);
    });
    test("import with external", () => {
      const src = `import "a" as b with external c`;
      treeScopeTestCase(src);
    });
    test("external", () => {
      const src = `external y`;
      treeScopeTestCase(src);
    });
    test("private declare", () => {
      const src = `z := y+1`;
      treeScopeTestCase(src);
    });
    test("public declare", () => {
      const src = `export x := z+123`;
      treeScopeTestCase(src);
    });
    test("export main", () => {
      const src = `export args -> {}`;
      treeScopeTestCase(src);
    });
    test("operator", () => {
      const src = `operator _+_ := fn x, y -> x + y`;
      treeScopeTestCase(src);
    });
    test("operator with precedence", () => {
      const src = `operator _+_ precedence 1 := fn x, y -> x + y`;
      treeScopeTestCase(src);
    });
    test("operator with tuple precedence", () => {
      const src = `operator _+_ precedence 1, 2 := fn x, y -> x + y`;
      treeScopeTestCase(src);
    });
  });
});

describe.todo("examples", () => {
  describe("procedural", () => {
    test("hello world", () => {
      const src = `
        external println
        export args -> {
          println "Hello, World!"
        }`;
      exampleScopeTestCase(src);
    });

    test("fibonacci", () => {
      const src = `
        export fib := n -> {
          if n < 2
            1
          else
            fib (n - 1) + fib (n - 2)
        }`;

      exampleScopeTestCase(src);
    });

    test.todo("quick sort recursive", () => {
      const src = `
        export quicksort := xs -> {
          if xs == ()
            return ()
          
          (pivot, ...rest) := xs
          smaller := for x in rest: if x < pivot: x
          bigger := for x in rest: if x >= pivot: x
          return (...quicksort smaller, pivot, ...quicksort bigger)
        }`;

      exampleScopeTestCase(src);
    });

    test.todo("bubble sort imperative", () => {
      const src = `
        export bubblesort := fn mut list -> {
          for i in range 0 list.length
          for j in range 0 (list.length - i - 1)
          if list[j] > list[j+1]
            list[j], list[j+1] = list[j+1], list[j]
          return list
        }`;

      exampleScopeTestCase(src);
    });
  });

  describe.todo("functional", () => {
    test("option", () => {
      const src = `
        export none := fn some, none -> none
        export some := value -> fn some, none -> some value
        export map := fn f, option -> match option {
          none -> none
          some value -> some (f value)
        }
        export unwrap := fn option, default -> match option {
          none -> default
          some value -> value
        }
        export flat_map := fn f, option -> match option {
          none -> none
          some value -> f value
        }
      `;
      exampleScopeTestCase(src);
    });

    test("result", () => {
      const src = `
        export ok := value -> fn ok, err -> ok value
        export err := value -> fn ok, err -> err value
        export map := fn f, result ->
          if result is ok value: ok (f value) else: result
        export unwrap := fn result, default ->
          if result is ok value: value else: default
        export flat_map := fn f, result ->
          if result is ok value: f value else: result
      `;
      exampleScopeTestCase(src);
    });

    test("tuple", () => {
      const src = `
        export tuple := fn x, y -> fn match -> match x y
        export first := fn tuple -> tuple (fn x, y -> x)
        export second := fn tuple -> tuple (fn x, y -> y)
      `;
      exampleScopeTestCase(src);
    });

    test("list", () => {
      const src = `
        symbol EmptyList
        export nil := fn cons, nil -> nil
        export cons := fn x, xs -> fn cons, nil -> cons x xs
        export head := fn list -> list (fn x, xs -> x) EmptyList
        export tail := fn list -> list (fn x, xs -> xs) EmptyList
        export map := fn f, list -> list (fn x, xs -> cons (f x) (map f xs)) nil
      `;
      exampleScopeTestCase(src);
    });

    test.todo("map", () => {
      const src = `
        symbol NotFound
        export empty := fn entry, empty -> empty
        export entry := fn key, value, rest -> fn entry, empty -> entry key value rest
        export find := fn key, map -> map
          (fn _key, value, rest ->
            if key == _key: value else find key rest
          )
          NotFound
        export insert := fn key, value, map -> entry key value (remove key map)
        export remove := fn key, map -> map
          (fn _key, value, rest ->
            if key == _key: rest else remove key rest
          )
          empty
        operator _[_] = fn map, key -> find key map
      `;
      exampleScopeTestCase(src);
    });
  });

  test.todo("prototype", () => {
    const src = `
      symbol Prototype
      operator _._ = fn value, name -> 
        if value[name] is some value: value
        else accessor value[Prototype] name
    `;
    exampleScopeTestCase(src);
  });
});
