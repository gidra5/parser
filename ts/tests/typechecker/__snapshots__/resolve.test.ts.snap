// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`expressions > function expressions > function 1`] = `
{
  "children": [
    {
      "children": [],
      "data": {
        "scope": Scope {
          "names": {},
          "scope": [],
        },
      },
      "name": "name",
      "value": "x",
    },
    {
      "children": [],
      "data": {
        "relativeIndex": 0,
        "scope": Scope {
          "names": {
            "x": 0,
          },
          "scope": [
            {
              "name": "x",
              "value": {},
            },
          ],
        },
      },
      "name": "name",
      "value": "x",
    },
  ],
  "data": {
    "scope": Scope {
      "names": {},
      "scope": [],
    },
  },
  "name": "operator",
  "value": "->",
}
`;

exports[`expressions > function expressions > function multiple params 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "data": {
            "scope": Scope {
              "names": {},
              "scope": [],
            },
          },
          "name": "name",
          "value": "x",
        },
        {
          "children": [],
          "data": {
            "scope": Scope {
              "names": {
                "x": 0,
              },
              "scope": [
                {
                  "name": "x",
                  "value": {},
                },
              ],
            },
          },
          "name": "name",
          "value": "y",
        },
      ],
      "data": {
        "scope": Scope {
          "names": {},
          "scope": [],
        },
      },
      "name": "operator",
      "value": "application",
    },
    {
      "children": [
        {
          "children": [],
          "data": {
            "relativeIndex": 1,
            "scope": Scope {
              "names": {
                "x": 0,
                "y": 1,
              },
              "scope": [
                {
                  "name": "x",
                  "value": {},
                },
                {
                  "name": "y",
                  "value": {},
                },
              ],
            },
          },
          "name": "name",
          "value": "x",
        },
        {
          "children": [],
          "data": {
            "relativeIndex": 0,
            "scope": Scope {
              "names": {
                "x": 0,
                "y": 1,
              },
              "scope": [
                {
                  "name": "x",
                  "value": {},
                },
                {
                  "name": "y",
                  "value": {},
                },
              ],
            },
          },
          "name": "name",
          "value": "y",
        },
      ],
      "data": {
        "scope": Scope {
          "names": {
            "x": 0,
            "y": 1,
          },
          "scope": [
            {
              "name": "x",
              "value": {},
            },
            {
              "name": "y",
              "value": {},
            },
          ],
        },
      },
      "name": "operator",
      "value": "application",
    },
  ],
  "data": {
    "scope": Scope {
      "names": {},
      "scope": [],
    },
  },
  "name": "operator",
  "value": "fn",
}
`;

exports[`expressions > function expressions > function with deep shadowed name access 1`] = `
{
  "children": [
    {
      "children": [],
      "data": {
        "scope": Scope {
          "names": {},
          "scope": [],
        },
      },
      "name": "name",
      "value": "a",
    },
    {
      "children": [
        {
          "children": [],
          "data": {
            "scope": Scope {
              "names": {
                "a": 0,
              },
              "scope": [
                {
                  "name": "a",
                  "value": {},
                },
              ],
            },
          },
          "name": "name",
          "value": "a",
        },
        {
          "children": [
            {
              "children": [],
              "data": {
                "scope": Scope {
                  "names": {
                    "a": 1,
                  },
                  "scope": [
                    {
                      "name": "a",
                      "value": {},
                    },
                    {
                      "name": "a",
                      "value": {},
                    },
                  ],
                },
              },
              "name": "name",
              "value": "a",
            },
            {
              "children": [
                {
                  "children": [
                    {
                      "children": [],
                      "data": {
                        "relativeIndex": 0,
                        "scope": Scope {
                          "names": {
                            "a": 2,
                          },
                          "scope": [
                            {
                              "name": "a",
                              "value": {},
                            },
                            {
                              "name": "a",
                              "value": {},
                            },
                            {
                              "name": "a",
                              "value": {},
                            },
                          ],
                        },
                      },
                      "name": "name",
                      "value": "a",
                    },
                  ],
                  "data": {
                    "relativeIndex": 1,
                    "scope": Scope {
                      "names": {
                        "a": 2,
                      },
                      "scope": [
                        {
                          "name": "a",
                          "value": {},
                        },
                        {
                          "name": "a",
                          "value": {},
                        },
                        {
                          "name": "a",
                          "value": {},
                        },
                      ],
                    },
                  },
                  "name": "operator",
                  "value": "#",
                },
              ],
              "data": {
                "relativeIndex": 2,
                "scope": Scope {
                  "names": {
                    "a": 2,
                  },
                  "scope": [
                    {
                      "name": "a",
                      "value": {},
                    },
                    {
                      "name": "a",
                      "value": {},
                    },
                    {
                      "name": "a",
                      "value": {},
                    },
                  ],
                },
              },
              "name": "operator",
              "value": "#",
            },
          ],
          "data": {
            "scope": Scope {
              "names": {
                "a": 1,
              },
              "scope": [
                {
                  "name": "a",
                  "value": {},
                },
                {
                  "name": "a",
                  "value": {},
                },
              ],
            },
          },
          "name": "operator",
          "value": "fn",
        },
      ],
      "data": {
        "scope": Scope {
          "names": {
            "a": 0,
          },
          "scope": [
            {
              "name": "a",
              "value": {},
            },
          ],
        },
      },
      "name": "operator",
      "value": "fn",
    },
  ],
  "data": {
    "scope": Scope {
      "names": {},
      "scope": [],
    },
  },
  "name": "operator",
  "value": "fn",
}
`;

exports[`expressions > function expressions > function with name shadowing 1`] = `
{
  "children": [
    {
      "children": [],
      "data": {
        "scope": Scope {
          "names": {},
          "scope": [],
        },
      },
      "name": "name",
      "value": "a",
    },
    {
      "children": [
        {
          "children": [],
          "data": {
            "scope": Scope {
              "names": {
                "a": 0,
              },
              "scope": [
                {
                  "name": "a",
                  "value": {},
                },
              ],
            },
          },
          "name": "name",
          "value": "a",
        },
        {
          "children": [],
          "data": {
            "relativeIndex": 0,
            "scope": Scope {
              "names": {
                "a": 1,
              },
              "scope": [
                {
                  "name": "a",
                  "value": {},
                },
                {
                  "name": "a",
                  "value": {},
                },
              ],
            },
          },
          "name": "name",
          "value": "a",
        },
      ],
      "data": {
        "scope": Scope {
          "names": {
            "a": 0,
          },
          "scope": [
            {
              "name": "a",
              "value": {},
            },
          ],
        },
      },
      "name": "operator",
      "value": "fn",
    },
  ],
  "data": {
    "scope": Scope {
      "names": {},
      "scope": [],
    },
  },
  "name": "operator",
  "value": "fn",
}
`;

exports[`expressions > function expressions > function with no arg 1`] = `
{
  "children": [
    {
      "children": [],
      "data": {
        "scope": Scope {
          "names": {},
          "scope": [],
        },
      },
      "name": "placeholder",
    },
    {
      "children": [
        {
          "children": [],
          "data": {},
          "name": "int",
          "value": 0,
        },
      ],
      "data": {
        "relativeIndex": 0,
        "scope": Scope {
          "names": {},
          "scope": [
            {
              "value": {},
            },
          ],
        },
      },
      "name": "operator",
      "value": "#",
    },
  ],
  "data": {
    "scope": Scope {
      "names": {},
      "scope": [],
    },
  },
  "name": "operator",
  "value": "fn",
}
`;

exports[`expressions > function expressions > function with placeholder arg 1`] = `
{
  "children": [
    {
      "children": [],
      "data": {
        "scope": Scope {
          "names": {},
          "scope": [],
        },
      },
      "name": "placeholder",
    },
    {
      "children": [
        {
          "children": [],
          "data": {},
          "name": "int",
          "value": 0,
        },
      ],
      "data": {
        "relativeIndex": 0,
        "scope": Scope {
          "names": {},
          "scope": [
            {
              "value": {},
            },
          ],
        },
      },
      "name": "operator",
      "value": "#",
    },
  ],
  "data": {
    "scope": Scope {
      "names": {},
      "scope": [],
    },
  },
  "name": "operator",
  "value": "->",
}
`;

exports[`expressions > function expressions > function with shadowed name access 1`] = `
{
  "children": [
    {
      "children": [],
      "data": {
        "scope": Scope {
          "names": {},
          "scope": [],
        },
      },
      "name": "name",
      "value": "a",
    },
    {
      "children": [
        {
          "children": [],
          "data": {
            "scope": Scope {
              "names": {
                "a": 0,
              },
              "scope": [
                {
                  "name": "a",
                  "value": {},
                },
              ],
            },
          },
          "name": "name",
          "value": "a",
        },
        {
          "children": [
            {
              "children": [],
              "data": {
                "relativeIndex": 0,
                "scope": Scope {
                  "names": {
                    "a": 1,
                  },
                  "scope": [
                    {
                      "name": "a",
                      "value": {},
                    },
                    {
                      "name": "a",
                      "value": {},
                    },
                  ],
                },
              },
              "name": "name",
              "value": "a",
            },
          ],
          "data": {
            "relativeIndex": 1,
            "scope": Scope {
              "names": {
                "a": 1,
              },
              "scope": [
                {
                  "name": "a",
                  "value": {},
                },
                {
                  "name": "a",
                  "value": {},
                },
              ],
            },
          },
          "name": "operator",
          "value": "#",
        },
      ],
      "data": {
        "scope": Scope {
          "names": {
            "a": 0,
          },
          "scope": [
            {
              "name": "a",
              "value": {},
            },
          ],
        },
      },
      "name": "operator",
      "value": "fn",
    },
  ],
  "data": {
    "scope": Scope {
      "names": {},
      "scope": [],
    },
  },
  "name": "operator",
  "value": "fn",
}
`;

exports[`expressions > pattern matching > in function parameters 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "data": {
                "scope": Scope {
                  "names": {},
                  "scope": [],
                },
              },
              "name": "name",
              "value": "x",
            },
            {
              "children": [],
              "data": {
                "scope": Scope {
                  "names": {
                    "x": 0,
                  },
                  "scope": [
                    {
                      "name": "x",
                      "value": {},
                    },
                  ],
                },
              },
              "name": "name",
              "value": "y",
            },
          ],
          "data": {
            "scope": Scope {
              "names": {},
              "scope": [],
            },
          },
          "name": "operator",
          "value": ",",
        },
      ],
      "data": {
        "scope": Scope {
          "names": {},
          "scope": [],
        },
      },
      "name": "group",
      "value": "parens",
    },
    {
      "children": [
        {
          "children": [],
          "data": {
            "relativeIndex": 1,
            "scope": Scope {
              "names": {
                "x": 0,
                "y": 1,
              },
              "scope": [
                {
                  "name": "x",
                  "value": {},
                },
                {
                  "name": "y",
                  "value": {},
                },
              ],
            },
          },
          "name": "name",
          "value": "x",
        },
        {
          "children": [],
          "data": {
            "relativeIndex": 0,
            "scope": Scope {
              "names": {
                "x": 0,
                "y": 1,
              },
              "scope": [
                {
                  "name": "x",
                  "value": {},
                },
                {
                  "name": "y",
                  "value": {},
                },
              ],
            },
          },
          "name": "name",
          "value": "y",
        },
      ],
      "data": {
        "scope": Scope {
          "names": {
            "x": 0,
            "y": 1,
          },
          "scope": [
            {
              "name": "x",
              "value": {},
            },
            {
              "name": "y",
              "value": {},
            },
          ],
        },
      },
      "name": "operator",
      "value": "+",
    },
  ],
  "data": {
    "scope": Scope {
      "names": {},
      "scope": [],
    },
  },
  "name": "operator",
  "value": "->",
}
`;
